src/promql/src/planner.rs

```diff
@@ -0,0 +1,627 @@
+        let res = match &prom_expr {
+            PromExpr::AggregateExpr { .. } => UnsupportedExprSnafu {
+                name: "Prom Aggregate",
+            }
+            .fail()?,
+            PromExpr::UnaryExpr { .. } => UnsupportedExprSnafu {
+                name: "Prom Unary Expr",
+            }
+            .fail()?,
+            PromExpr::BinaryExpr { lhs, rhs, .. } => {
+                let _left_input = self.prom_expr_to_plan(*lhs.clone())?;
+                let _right_input = self.prom_expr_to_plan(*rhs.clone())?;
+
+                UnsupportedExprSnafu {
+                    name: "Prom Binary Expr",
+                }
+                .fail()?
+            }
+            PromExpr::ParenExpr { .. } => UnsupportedExprSnafu {
+                name: "Prom Paren Expr",
+            }
+            .fail()?,
+            PromExpr::SubqueryExpr { .. } => UnsupportedExprSnafu {
+                name: "Prom Subquery",
+            }
+            .fail()?,
+            PromExpr::NumberLiteral { .. } => UnsupportedExprSnafu {
+                name: "Prom Aggregate",
+            }
+            .fail()?,
+            PromExpr::StringLiteral { .. } => UnsupportedExprSnafu {
+                name: "Prom Aggregate",
+            }
+            .fail()?,
+            PromExpr::VectorSelector {
+                name,
+                offset,
+                start_or_end: _,
+                label_matchers,
+            } => {
+                // This `name` should not be optional
+                let name = name.as_ref().unwrap().clone();
+                self.setup_context(&name)?;
+                let normalize =
+                    self.selector_to_series_normalize_plan(&name, *offset, label_matchers.clone())?;
+                let manipulate = InstantManipulate::new(
+                    self.ctx.start,
+                    self.ctx.end,
+                    self.ctx.lookback_delta,
+                    self.ctx.interval,
+                    self.ctx
+                        .time_index_column
+                        .clone()
+                        .with_context(|| TimeIndexNotFoundSnafu { table: name })?,
+                    normalize,
+                );
+                LogicalPlan::Extension(Extension {
+                    node: Arc::new(manipulate),
+                })
+            }
+            PromExpr::MatrixSelector { .. } => UnsupportedExprSnafu {
+                name: "Prom Matrix Selector",
+            }
+            .fail()?,
+            PromExpr::Call { func, args } => {
+                let args = self.create_function_args(args)?;
+                let input =
+                    self.prom_expr_to_plan(args.input.with_context(|| ExpectExprSnafu {
+                        expr: prom_expr.clone(),
+                    })?)?;
+                let mut func_exprs = self.create_function_expr(func, args.literals)?;
+                func_exprs.insert(0, self.create_time_index_column_expr()?);
```

Comment:
```suggestion
                // todo: reserve capacity first
                let mut func_exprs = vec![self.create_time_index_column_expr()?];
                self.create_function_expr(func, args.literals, &mut func_exprs)?;
```

Reply:
Let's keep it simple and clear (your rewriting is wrong). It's not very necessary to reserve this one capacity.